### PR TITLE
Fix None grad problem during training TOOD by adding SigmoidGeometricMean

### PR DESCRIPTION
**Pull request by Johnson-Wang**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

The training of TOOD often encounters None gradient during backpropagation, which would further cause None tensors in the next training step. Some issues in the original repo (https://github.com/fcjian/TOOD/issues/11) might be also due to this error. The problem is caused by the naive implementation of sigmoid geometric mean function ``cls_score = (cls_logits.sigmoid() * cls_prob.sigmoid()).sqrt()``. This output might be 0 if  ``cls_logits`` or ``cls_prob`` is a low negative value, which causes either inf grad of none grad during backpropagation. 

## Modification

A reimplementation of ``SigmoidGeometricMean`` class as an inheritance of ``torch.autograd.Function`` is proposed. The backward function is derived analytically and would avoid and inf or none grad during bp. 
- This modification has little influence on the final results (42.3 mAP after modification vs. 42.4 mAP as reported).
- This modification enables users to train TOOD without ATSS warmup, yet with some performance drop (~41.8 mAP)

## Checklist

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] This PR does not involve any function interface change.
- [x] Docstring has been added.

### CODE REVIEW

**Review by ZwwWayne — APPROVED** [review]

File: mmdet/models/utils/misc.py [hunk 1]
```diff
@@ -1,7 +1,34 @@
 # Copyright (c) OpenMMLab. All rights reserved.
+from torch.autograd import Function
 from torch.nn import functional as F
 
 
+class SigmoidGeometricMean(Function):
```
Comment (ZwwWayne) [inline mmdet/models/utils/misc.py:6]: How about we implement an interface named sigmoid_geometric_mean = SigmoidGeometricMean.apply here so that in tood_head we can simply use sigmoid_geometric_mean(xxx)?

Comment (jshilong) [reply mmdet/models/utils/misc.py:6]: We have to use `SigmoidGeometricMean.apply` in the high PyTorch version such as  Pytorch 1.10

**Review by jshilong — APPROVED** [review]
LGTM

**Review by RangiLyu — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]
